amethyst_core/src/transform/components/local_transform.rs

```diff
@@ -21,124 +20,186 @@ pub struct Transform {
+        let x = Vector4 {
+            x: quat.x.x * self.scale.x,
+            y: quat.x.y * self.scale.y,
+            z: quat.x.z * self.scale.z,
+            w: 0.0
+        };
+        let y = Vector4 {
+            x: quat.y.x * self.scale.x,
+            y: quat.y.y * self.scale.y,
+            z: quat.y.z * self.scale.z,
+            w: 0.0
+        };
+        let z = Vector4 {
+            x: quat.z.x * self.scale.x,
+            y: quat.z.y * self.scale.y,
+            z: quat.z.z * self.scale.z,
+            w: 0.0
+        };
+
+        let mat = Matrix4 { x, y, z, w: self.translation.extend(1.0)};
+        mat
     }
 
-    /// Move relatively to its current position, but independently from its orientation.
-    /// Ideally, first normalize the direction and then multiply it
-    /// by whatever amount you want to move before passing the vector to this method
+    /// Move relatively to its current position.
     #[inline]
-    pub fn move_global(&mut self, direction: Vector3<f32>) -> &mut Self {
-        self.translation = self.translation + direction;
+    pub fn move_global(&mut self, translation: Vector3<f32>) -> &mut Self {
+        self.translation += translation;
         self
     }
 
     /// Move relatively to its current position and orientation.
+    ///
+    /// Equivalent to rotating the translation before applying.
     #[inline]
-    pub fn move_local(&mut self, axis: Vector3<f32>, amount: f32) -> &mut Self {
-        if axis.magnitude2() == 0.0 {
-            return self;
-        }
+    pub fn move_local(&mut self, translation: Vector3<f32>) -> &mut Self {
+        self.translation += self.rotation * translation;
+        self
+    }
 
-        let delta = Quaternion::from(self.rotation) * axis.normalize() * amount;
+    /// Move a distance along an axis.
+    ///
+    /// Will not move in the case where the axis is zero, for any distance.
+    #[inline]
+    pub fn move_along_global(&mut self, direction: Vector3<f32>, distance: f32) -> &mut Self {
+        if !ulps_eq!(direction, Zero::zero()) {
+            self.translation += direction.normalize() * distance;
+        }
+        self
+    }
 
-        self.translation = self.translation + delta;
+    /// Move a distance along an axis.
+    ///
+    /// Will not move in the case where the axis is zero, for any distance.
+    #[inline]
+    pub fn move_along_local(&mut self, direction: Vector3<f32>, distance: f32) -> &mut Self {
+        if !ulps_eq!(direction, Zero::zero()) {
+            self.translation += self.rotation * direction.normalize() * distance;
+        }
         self
     }
 
     /// Move forward relative to current position and orientation.
-    pub fn move_forward(&mut self, orientation: &Orientation, amount: f32) -> &mut Self {
-        self.move_local(orientation.forward.into(), amount)
+    #[inline]
+    pub fn move_forward(&mut self, amount: f32) -> &mut Self {
+        // sign is reversed because z comes towards us
```

Comment:
I thought it was the inverse?
Anyway, after this change the fly_camera forward would be inverted.

Reply:
I matched the OpenGL orientation (in the left-hand right-hand sense). I should check the shaders to see which orientation we use. If you run the fly camera example now it should all work -> if it doesn't that's a bug in my PR.

Reply:
:+1: